File: src/test/java/com/technews/crawling/OracleJavaBlogCrawlingTest.java
package com.technews.crawling;

import lombok.*;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.util.*;

@Slf4j
public class OracleJavaBlogCrawlingTest {

    private static final String BLOG_BASE_URL = "https://blogs.oracle.com/java/";
    private static final List<String> CATEGORIES = List.of("Product & Ecosystem", "Java Technology");

    private static Set<String> SAVED_POST_TITLE = new HashSet<>();

    @Test
    void test() {
        final List<OracleJavaBlogPost> posts = getPosts();
        System.out.println("posts.size() = " + posts.size());
        posts.forEach(System.out::println);
    }

    private static List<OracleJavaBlogPost> getPosts() {
        final List<OracleJavaBlogPost> result = new ArrayList<>();

        try {
            final Document doc = Jsoup.connect(BLOG_BASE_URL)
                    .header("User-Agent", "PostmanRuntime/7.37.3")
                    .get();
            final Elements categoryElements = doc.select(".with-category");
            for (Element categoryElement : categoryElements) {
                final String category = categoryElement.select(".rw-ptitle").text();
                if (CATEGORIES.contains(category)) {
                    result.addAll(getPosts(category, categoryElement));
                }
            }
        } catch (Exception e) {
            log.error("jsoup parsing exception. message: {}", e.getMessage(), e);
            return Collections.EMPTY_LIST;
        }

        return result;
    }

    private static List<OracleJavaBlogPost> getPosts(final String category, final Element postElement) {
        final List<OracleJavaBlogPost> result = new ArrayList<>();

        final Elements select = postElement.select(".cscroll-item-w1");
        for (final Element element : select) {
            final PostInfo postInfo = getPostInfo(element);
            if (SAVED_POST_TITLE.contains(postInfo.title())) {
                continue;
            }

            final OracleJavaBlogPost post = OracleJavaBlogPost.builder()
                    .subject("java")
                    .title(postInfo.title())
                    .url(postInfo.url())
                    .category("oracle")
                    .writer(postInfo.writer())
                    .date(LocalDate.now().toString())
                    .tags(List.of("oracle", category))
                    .createdDt(LocalDate.now().toString())
                    .build();

            result.add(post);
            SAVED_POST_TITLE.add(postInfo.title());
        }

        return result;
    }

    private static PostInfo getPostInfo(final Element element) {
        try {
            final Elements blogTile = element.select(".blogtile-w2");
            final Elements post = blogTile.select("a");
            return PostInfo.builder()
                    .title(post.get(2).text())
                    .url(BLOG_BASE_URL + post.get(2).attr("href"))
                    .writer(post.get(3).text())
                    .build();
        } catch (Exception e) {
            return PostInfo.EMPTY;
        }
    }

    @Getter
    @Builder
    @ToString
    @NoArgsConstructor
    @AllArgsConstructor
    private static class OracleJavaBlogPost {
        private String subject;
        private String title;
        private String url;
        private String category;
        private String writer;
        private String date;
        List<String> tags;
        private String createdDt;

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            OracleJavaBlogPost post = (OracleJavaBlogPost) o;
            return title == post.title;
        }

        @Override
        public int hashCode() {
            return Objects.hash(title);
        }
    }

    private record PostInfo(
            String title,
            String url,
            String writer
    ) {
        public static final PostInfo EMPTY = new PostInfo(StringUtils.EMPTY, StringUtils.EMPTY, StringUtils.EMPTY);

        @Builder
        private PostInfo {
        }
    }
}

File: src/test/java/com/technews/crawling/OracleJavaBlogCrawlingTest.kt
package com.technews.crawling

import io.kotest.core.spec.style.BehaviorSpec
import mu.KotlinLogging
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.time.LocalDate

private val logger = KotlinLogging.logger {}

class OracleJavaBlogCrawlingTest : BehaviorSpec({
    Given("Oracle Java Blog 정보 크롤링") {
        val posts = fetchPosts()
        println("posts.size = ${posts.size}")
        posts.forEach { println(it) }
    }
}) {
    companion object {
        private const val BLOG_BASE_URL = "https://blogs.oracle.com/java/"
        private val CATEGORIES = listOf("Product & Ecosystem", "Java Technology")
        private val SAVED_POST_TITLE: MutableSet<String> = mutableSetOf()

        private fun fetchPosts(): List<OracleJavaBlogPost> {
            return runCatching {
                Jsoup.connect(BLOG_BASE_URL)
                    .header("User-Agent", "PostmanRuntime/7.37.3")
                    .get()
                    .select(".with-category")
                    .flatMap { categoryElement ->
                        val category = categoryElement.select(".rw-ptitle").text()
                        if (category in CATEGORIES) {
                            getPosts(category, categoryElement)
                        } else {
                            emptyList()
                        }
                    }
            }.getOrElse {
                logger.error("Jsoup parsing exception. Message: ${it.message}", it)
                emptyList()
            }
        }

        private fun getPosts(category: String, postElement: Element): List<OracleJavaBlogPost> {
            return postElement.select(".cscroll-item-w1")
                .mapNotNull { element ->
                    val postInfo = getPostInfo(element)
                    if (SAVED_POST_TITLE.add(postInfo.title)) {
                        OracleJavaBlogPost(
                            subject = "java",
                            title = postInfo.title,
                            url = postInfo.url,
                            category = "oracle",
                            writer = postInfo.writer,
                            date = LocalDate.now().toString(),
                            tags = listOf("oracle", category),
                            createdDt = LocalDate.now().toString()
                        )
                    } else {
                        null
                    }
                }
        }

        private fun getPostInfo(element: Element): PostInfo {
            return runCatching {
                val blogTile = element.select(".blogtile-w2")
                val postLinks = blogTile.select("a")
                PostInfo(
                    title = postLinks.getOrNull(2)?.text().orEmpty(),
                    url = BLOG_BASE_URL + postLinks.getOrNull(2)?.attr("href").orEmpty(),
                    writer = postLinks.getOrNull(3)?.text().orEmpty()
                )
            }.getOrDefault(PostInfo())
        }

        private data class OracleJavaBlogPost(
            val subject: String = "",
            val title: String = "",
            val url: String = "",
            val category: String = "",
            val writer: String = "",
            val date: String = "",
            var tags: List<String> = emptyList(),
            val createdDt: String = "",
        )

        private data class PostInfo(
            val title: String = "",
            val url: String = "",
            val writer: String = "",
        )
    }
}
